catboost/private/libs/algo/greedy_tensor_search.cpp

```diff
@@ -1393,62 +1564,133 @@ static TNonSymmetricTreeStructure GreedyTensorSearchDepthwise(
         TVector<TIndexType> splittedLeafs;
         TVector<TIndexType> nextLevelLeafs;
         const size_t maxFeatureValueCount = CalcMaxFeatureValueCount(*fold, candidatesContexts);
-        for (TIndexType leafToSplit : curLevelLeafs) {
-            const auto& leafBounds = ctx->SampledDocs.LeavesBounds[leafToSplit];
-            if (leafBounds.GetSize() < ctx->Params.ObliviousTreeOptions->MinDataInLeaf) {
+        const ui64 statsSize = maxBucketCount * nFeatures * maxSplitEnsembles;
+        const bool isMultiClass = fold->GetApproxDimension() != 1;
+
+        TSubtarctTrickInfo subTrickInfo(
+            &data,
+            &candidatesContexts,
+            fold,
+            ctx,
+            &parentsQueue,
+            isMultiClass,
+            scoreStDev,
+            maxBucketCount,
+            maxSplitEnsembles,
+            statsSize,
+            maxFeatureValueCount);
```

Comment:
0. let us create candidatesContexts and calculate maxBucketCount, maxSplitEnsembles, scoreStDev, statsSize, and maxFeatureValueCount in TSubtractTrickInfo's ctor or some helper method in TSubtractTrickInfo (move there lines 1535-1555)

1. after that TSubtractTrickInfo can be constructed from just &data, fold, ctx, and &parentsQueue

2. let us check for multi-class and multi-regression in leafwise_scoring.cpp at line 350, and add CB_ENSURE(approxDimension == 1, "Subtraction trick is not implemented for MultiClass and MultiRegression") here

3. it seems that we should multiply by approxDimension when we calculate statsSize? if yes, please add a comment.